PagingWithNetworkSample/lib/src/main/java/com/android/example/paging/pagingwithnetwork/reddit/paging/SynchronousRemoteStates.kt

```diff
@@ -0,0 +1,141 @@
+        }
+        computeSynchronousRemoteStates(
+            sourceRefreshState = combinedLoadStates.source.refresh,
+            sourceState = combinedLoadStates.source.prepend,
+            remoteState = combinedLoadStates.mediator?.prepend,
+            synchronousRemoteState = prependState,
+        ).also {
+            prepend = it.first
+            prependState = it.second
+        }
+        computeSynchronousRemoteStates(
+            sourceRefreshState = combinedLoadStates.source.refresh,
+            sourceState = combinedLoadStates.source.append,
+            remoteState = combinedLoadStates.mediator?.append,
+            synchronousRemoteState = appendState,
+        ).also {
+            append = it.first
+            appendState = it.second
+        }
+    }
+
+    private fun computeSynchronousRemoteStates(
+        sourceRefreshState: LoadState,
+        sourceState: LoadState,
+        remoteState: LoadState?,
+        synchronousRemoteState: SynchronousRemoteState,
+    ): Pair<LoadState, SynchronousRemoteState> {
+        if (remoteState == null) return sourceState to NOT_LOADING
+
+        return when (synchronousRemoteState) {
+            NOT_LOADING -> when (remoteState) {
+                is Loading -> Loading to REMOTE_STARTED
+                is Error -> remoteState to REMOTE_ERROR
+                else -> NotLoading(remoteState.endOfPaginationReached) to NOT_LOADING
+            }
+            REMOTE_STARTED -> when {
+                remoteState is Error -> remoteState to REMOTE_ERROR
+                sourceRefreshState is Loading -> Loading to SOURCE_LOADING
+                else -> Loading to REMOTE_STARTED
+            }
+            REMOTE_ERROR -> when (remoteState) {
+                is Error -> remoteState to REMOTE_ERROR
+                else -> Loading to REMOTE_STARTED
+            }
+            SOURCE_LOADING -> when {
+                sourceRefreshState is Error -> sourceRefreshState to SOURCE_ERROR
+                remoteState is Error -> remoteState to REMOTE_ERROR
+                sourceRefreshState is NotLoading -> {
+                    NotLoading(remoteState.endOfPaginationReached) to NOT_LOADING
+                }
+                else -> Loading to SOURCE_LOADING
+            }
+            SOURCE_ERROR -> when (sourceRefreshState) {
+                is Error -> sourceRefreshState to SOURCE_ERROR
+                else -> sourceRefreshState to SOURCE_LOADING
+            }
+        }
+    }
+}
+
+@OptIn(ExperimentalCoroutinesApi::class)
+fun Flow<CombinedLoadStates>.asSynchronousRemoteStates(): Flow<LoadStates> {
+    val syncRemoteState = SynchronousRemoteStates()
+    return scan(syncRemoteState.toLoadStates()) { _, combinedLoadStates ->
+        syncRemoteState.updateFromCombinedLoadStates(combinedLoadStates)
+        syncRemoteState.toLoadStates()
+    }
+}
+
+/**
+ * State machine used to compute [LoadState] values in [SynchronousRemoteStates].
+ */
+enum class SynchronousRemoteState {
```

Comment:
private, possibly move into SynchronousRemoteStates, since it's an impl detail.

Could call it "State" or something

Reply:
Done